packages/oidc/src/token.ts

```diff
@@ -0,0 +1,224 @@
+      }
+      dir = path.dirname(dir);
+    }
+  } catch (e) {
+    throw new VercelOidcTokenError(
+      'Token refresh only supported in node server environments'
+    );
+  }
+  throw new VercelOidcTokenError('Unable to find root directory');
+}
+
+function saveToken(token: VercelTokenResponse, projectId: string): void {
+  try {
+    const dir = getUserDataDir();
+    if (!dir) {
+      throw new VercelOidcTokenError('Unable to find user data directory');
+    }
+    const tokenPath = path.join(dir, 'com.vercel.token', `${projectId}.json`);
+    const tokenJson = JSON.stringify(token);
+    fs.mkdirSync(path.dirname(tokenPath), { recursive: true });
+    fs.writeFileSync(tokenPath, tokenJson);
+    fs.chmodSync(tokenPath, 0o600);
+    return;
+  } catch (e) {
+    throw new VercelOidcTokenError(`Failed to save token`, e);
+  }
+}
+
+function loadToken(projectId: string): VercelTokenResponse | null {
+  try {
+    const dir = getUserDataDir();
+    if (!dir) {
+      return null;
+    }
+    const tokenPath = path.join(dir, 'com.vercel.token', `${projectId}.json`);
+    if (!fs.existsSync(tokenPath)) {
+      return null;
+    }
+    const token = JSON.parse(fs.readFileSync(tokenPath, 'utf8'));
+    assertVercelOidcTokenResponse(token);
+    return token;
+  } catch (e) {
+    throw new VercelOidcTokenError(`Failed to load token`, e);
+  }
+}
+
+export function getTokenPayload(token: string): TokenPayload {
+  const tokenParts = token.split('.');
+  if (tokenParts.length !== 3) {
+    throw new VercelOidcTokenError('Invalid token');
+  }
+
+  const base64 = tokenParts[1].replace(/-/g, '+').replace(/_/g, '/');
+  const padded = base64.padEnd(
+    base64.length + ((4 - (base64.length % 4)) % 4),
+    '='
+  );
+  return JSON.parse(Buffer.from(padded, 'base64').toString('utf8'));
+}
+
+export function isExpired(token: TokenPayload): boolean {
+  const timeout = ms('15m');
+  return token.exp * 1000 < Date.now() + timeout;
+}
+
+export async function refreshToken(): Promise<void> {
+  let maybeToken = loadToken(findProjectId());
+  if (!maybeToken || isExpired(getTokenPayload(maybeToken.token))) {
+    const authToken = getVercelCliToken();
+    if (!authToken) {
+      throw new VercelOidcTokenError(
+        'Failed to refresh OIDC token: login to vercel cli'
+      );
+    }
+    const projectId = findProjectId();
+    if (!projectId) {
+      throw new VercelOidcTokenError(
+        'Failed to refresh OIDC token: project id not found'
+      );
+    }
+    maybeToken = await getVercelOidcToken(authToken, projectId);
+    if (!maybeToken) {
+      throw new VercelOidcTokenError('Failed to refresh OIDC token');
+    }
+    saveToken(maybeToken, projectId);
+  }
+  process.env.VERCEL_OIDC_TOKEN = maybeToken.token;
+  return;
+}
+
+export class VercelOidcTokenError extends Error {
+  cause?: unknown;
+
+  constructor(message: string, cause?: unknown) {
+    super(message);
+    this.name = 'VercelOidcTokenError';
+    this.cause = cause;
```

Comment:
I think you can do this? I.e., we don't have to implement `cause?: unknown` or override `toString` ourselves, do we?
```suggestion
    super(message, { cause });
    this.name = 'VercelOidcTokenError';
```

Reply:
The same ES2022 issue would apply here

Reply:
Ack, then let's leave as-is. Maybe it's best to just do the string prefixing 🤔